source/extensions/filters/http/grpc_json_transcoder/http_body_utils.cc

```diff
@@ -2,23 +2,70 @@
 
 #include "google/api/httpbody.pb.h"
 
+using Envoy::Protobuf::io::CodedInputStream;
 using Envoy::Protobuf::io::CodedOutputStream;
 using Envoy::Protobuf::io::StringOutputStream;
+using Envoy::Protobuf::io::ZeroCopyInputStream;
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace GrpcJsonTranscoder {
 
+namespace {
+
+// Embedded messages are treated the same way as strings (wire type 2).
+constexpr uint32_t ProtobufLengthDelimitedField = 2;
+
+bool parseMessageByFieldPath(CodedInputStream* input,
+                             absl::Span<const Protobuf::Field* const> field_path,
+                             Protobuf::Message* message) {
+  if (field_path.empty()) {
+    return message->MergeFromCodedStream(input);
+  }
+
+  const uint32_t expected_tag = (field_path.front()->number() << 3) | ProtobufLengthDelimitedField;
+  for (;;) {
+    const uint32_t tag = input->ReadTag();
+    if (tag == expected_tag) {
+      uint32_t length = 0;
+      if (!input->ReadVarint32(&length)) {
+        return false;
+      }
+      auto limit = input->IncrementRecursionDepthAndPushLimit(length);
+      if (!parseMessageByFieldPath(input, field_path.subspan(1), message)) {
+        return false;
+      }
+      if (!input->DecrementRecursionDepthAndPopLimit(limit.first)) {
+        return false;
+      }
+    } else if (tag == 0) {
+      return true;
+    } else {
+      if (!Protobuf::internal::WireFormatLite::SkipField(input, tag)) {
```

Comment:
Dependency on internal API worries me, but I didn't find a better way without actually copying the same logic.

Reply:
Sorry for the late response. Can we add more test coverage for this function? This part isn't covered.

https://344313-65214191-gh.circle-artifacts.com/0/coverage/coverage/proc/self/cwd/source/extensions/filters/http/grpc_json_transcoder/http_body_utils.cc.html#L20